fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java

```diff
@@ -872,6 +876,55 @@ protected OptExpression queryBasedRewrite(RewriteContext rewriteContext, ScalarO
+            PartitionInfo partitionInfo = mv.getPartitionInfo();
+            if (partitionInfo instanceof ExpressionRangePartitionInfo) {
+                ExpressionRangePartitionInfo expressionRangePartitionInfo = (ExpressionRangePartitionInfo) partitionInfo;
+                List<Column> partitionColumns = expressionRangePartitionInfo.getPartitionColumns();
+                Preconditions.checkState(partitionColumns.size() == 1);
+                ScalarOperator queryPredicate = rewriteContext.getQueryPredicateSplit().toScalarOperator();
+                List<ScalarOperator> queryPredicates = Utils.extractConjuncts(queryPredicate);
+                List<ScalarOperator> queryPartitionRelatedScalars = queryPredicates.stream()
+                        .filter(predicate -> isRelatedPredicate(predicate, partitionColumns.get(0).getName()))
+                        .collect(Collectors.toList());
+
+                ScalarOperator mvPredicate = rewriteContext.getMvPredicateSplit().toScalarOperator();
+                List<ScalarOperator> mvPredicates = Utils.extractConjuncts(mvPredicate);
+                List<ScalarOperator> mvPartitionRelatedScalars = mvPredicates.stream()
+                        .filter(predicate -> isRelatedPredicate(predicate, partitionColumns.get(0).getName()))
+                        .collect(Collectors.toList());
+                // when query has no partition related predicates and mv has,
+                // we should consider to add null value predicate into compensation predicates
+                ColumnRefOperator partitionColumnRef =
+                        getRelatedPredicate(mvPredicate, partitionColumns.get(0).getName());
+                if (queryPartitionRelatedScalars.isEmpty()
+                        && !mvPartitionRelatedScalars.isEmpty() && partitionColumnRef.isNullable()) {
+                    ColumnRewriter columnRewriter = new ColumnRewriter(rewriteContext);
+                    partitionColumnRef = columnRewriter.rewriteViewToQuery(partitionColumnRef).cast();
+                    IsNullPredicateOperator isNullPredicateOperator = new IsNullPredicateOperator(partitionColumnRef);
+                    IsNullPredicateOperator isNotNullPredicateOperator = new IsNullPredicateOperator(true, partitionColumnRef);
+                    List<ScalarOperator> predicates = Utils.extractConjuncts(queryCompensationPredicate);
+                    List<ScalarOperator> partitionRelatedPredicates = predicates.stream()
+                            .filter(predicate -> isRelatedPredicate(predicate, partitionColumns.get(0).getName()))
+                            .collect(Collectors.toList());
+                    predicates.removeAll(partitionRelatedPredicates);
+                    if (partitionRelatedPredicates.contains(isNullPredicateOperator)
+                            || partitionRelatedPredicates.contains(isNotNullPredicateOperator)) {
+                        if (partitionRelatedPredicates.size() != 1) {
```

Comment:
partititon column **is null**  and partititon column **is not null** ?

Reply:
when the compensation query predicate is isNotNullPredicateOperator, should add it back.